nuxeo-features/nuxeo-automation/nuxeo-automation-core/src/main/java/org/nuxeo/ecm/automation/core/operations/document/RemoveItemFromListProperty.java

```diff
@@ -64,43 +66,44 @@
     protected boolean save = true;
 
     @OperationMethod(collector = DocumentModelCollector.class)
-    public DocumentModel run(DocumentModel doc) throws OperationException, IOException {
-
-        if (index != null) { // clear just the specific property
-            Property complexProperty = doc.getProperty(xpath);
-            ListType listType = (ListType) complexProperty.getField().getType();
-
-            if (!listType.getFieldType().isComplexType() && !listType.isListType()) {
-                throw new OperationException("Property type is not supported by this operation");
-            }
-
-            ListProperty listProperty = (ListProperty) complexProperty;
-            List<?> propertiesValues = (List<?>) listProperty.getValue();
-            // remove the desired property
-            propertiesValues.remove(index.intValue());
-
-            listProperty.clear();
-            // set the remaining properties
-            listProperty.setValue(propertiesValues);
-
-        } else { // clear all the properties
-            Property complexProperty = doc.getProperty(xpath);
-            ListType listType = (ListType) complexProperty.getField().getType();
-
-            if (!listType.getFieldType().isComplexType() && !listType.isListType()) {
-                throw new OperationException("Property type is not supported by this operation");
-            }
+    public DocumentModel run(DocumentModel doc) throws OperationException {
+        Property property = doc.getProperty(xpath);
+        if (!property.isList()) {
+            throw new OperationException("Property type is not supported by this operation");
+        }
 
-            ListProperty listProperty = (ListProperty) complexProperty;
-            listProperty.clear();
+        ListType listType = (ListType) property.getType();
+        if (listType.isArray()) {
+            removeItemFromArrayProperty(doc, property);
+        } else {
+            removeItemFromListProperty(property);
         }
 
         doc = session.saveDocument(doc);
         if (save) {
             session.save();
         }
         return doc;
+    }
 
+    protected void removeItemFromArrayProperty(DocumentModel doc, Property property) {
+        if (index != null) {
+            Serializable[] value = (Serializable[]) property.getValue();
+            List<Serializable> list = new ArrayList<>(Arrays.asList(value));
+            list.remove(index.intValue());
+            doc.setPropertyValue(xpath, list.toArray(new Serializable[0]));
```

Comment:
Could you check just with `doc.setPropertyValue(xpath, list)`? I think the low level code does the conversion automatically to the expected type.

Reply:
It does, see https://github.com/nuxeo/nuxeo/blob/aa44238df08ccc9c1867888bdc7a3e7575c8d985/nuxeo-core/nuxeo-core-api/src/main/java/org/nuxeo/ecm/core/api/model/impl/ArrayProperty.java#L108